IBC.md

```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
 }
 ```
 
-TODO: explain how to handle/parse errors (As part of
-https://github.com/CosmWasm/cosmwasm/issues/762)
+##### Acknowledgement Processing
+
+A major issue that is unique to `ibc_packet_receive` is that it is expected to
+often reject an incoming packet, but it cannot abort the transaction. We
+actually expect all state changes from the contract (as well as dispatched
+messages) to be reverted when the packet is rejected, but the transaction to
+properly commit an acknowledgement with encoded error. In other words, this "IBC
+Handler" will error and revert, but the "IBC Router" must succeed and commit an
+acknowledgement message (that can be parsed by the sending chain as an error).
+
+The atomicity issue was first
+[analyzed in the Cosmos SDK implementation](https://github.com/cosmos/ibc-go/issues/68)
+and refined into
+[changing semantics of the OnRecvPacket SDK method](https://github.com/cosmos/ibc-go/issues/91),
+which was
+[implemented in April 2021](https://github.com/cosmos/ibc-go/pull/107), likely
+to be released with Cosmos SDK 0.43 or 0.44. Since we want the best,
+future-proof interface for contracts, we will use an approach inspired by that
+work, and add an adapter in `wasmd` until we can upgrade to a Cosmos SDK version
+that implements this.
+
+After quite some
+[discussion on how to encode the errors](https://github.com/CosmWasm/cosmwasm/issues/762),
+we identified a number of issues implementing this idea with the CosmWasm model:
+
+1. We only return `Ok(Response)` or `Err(String)`. There is no way to return an
+   arbitrary binary encoded message in the error variant.
+2. If we return `messages` to be dispatched, they may fail with an error, which
+   is returned directly to the x/wasm handler. Obviously an error from the bank
+   handler will not be in the proper IBC Acknowledgement format.
+3. We also realized that in some cases, the original `ibc_packet_receive` will
+   not even have the information needed to format the success response, so we
+   will need some delayed response-builder in this case also.
+
+eg. In the ICS27 spec, they define a packet to register a new account, the
+acknowledgement of that packet should contain the address of the new account on
+the receiving chain. If we implemented this in CosmWasm, the contract would have
+to return a submessage of `WasmMsg::Instantiate{}` and only get the proper
+address for the response in `reply`. Thus, we need to have some way to fully
+build the success acknowledgement after the fact as well.
+
+With this in mind, we designed the following workflow for the `x/wasm` runtime:
+
+1. Wrap the storage with a "local cache", so we don't immediately write out
+   changes
+2. Call `ibc_receive_packet`, get either `Err(String)` or
+   `Ok(Response + acknowledgement)`. If `Err`, jump to 4.
+3. Execute any `submessages` and `messages` returned with this response. If any
+   return `Err`, jump to 4.
```

Comment:
Do we really want to change how messages behave? I'd avoid that

Reply:
This is also not clear to me. Why do we need step 4 for messages? Submessages are available instead.

Reply:
`on_packet_receive` may return an error (revert contract state)

`OnPacketReceive` in `x/wasm` should *never* return an error, but rather revert the contract state and return binary data to the sender to explain it failed.

If my contract sends tokens, but for some reason that fails (insufficient funds), it should not abort the tx (so the same packet is posted again and again), but make no state change and return an error acknowledgement.

I suggest reading through the 3 linked cosmos-sdk/ibc-go issues/PRs and all the comments for full context. (I know it is long, but this is the approach they are going for in the sdk and what I want to support in CosmWasm)

Reply:
In go terms, the `ibc_encode_acknowledgement` works like a `defer()` on  the result

Reply:
If you understand what I am trying to explain I would be grateful for a better way to express it